File: src/com/cattailsw/nanidroid/SurfaceManager.java
package com.cattailsw.nanidroid;

import java.util.HashMap;
import java.util.Map;
import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import java.util.Set;

public class SurfaceManager {

    private static SurfaceManager _self = null;
    protected SurfaceManager() {

    }

    public static SurfaceManager getInstance() {
	if ( _self == null )
	    _self = new SurfaceManager();

	return _self;
    }

    Map<String, ShellSurface> surfaces = null;
    Map<String, String[]> sakuraAliasTable = null; // alias can be N-to-one?
    Map<String, String[]> keroAliasTable = null;

    public int addSurface(String id, ShellSurface s) {
	if ( surfaces == null )
	    surfaces = new HashMap<String, ShellSurface>();

	surfaces.put(id, s);

	return surfaces.size();
    }

    public boolean containsSurface(String id) {
	if ( surfaces == null )
	    return false;
	return surfaces.containsKey(id);
    }

    public int getTotalSurfaceCount() {
	return surfaces.size();
    }

    public Set<String> getSurfaceKeys() {
	return surfaces.keySet();
    }

    public ShellSurface getSakuraSurface(String id) {
	if ( surfaces.containsKey(id) )
	return surfaces.get(id);
	else
	    return surfaces.get("0");
    }

    public ShellSurface getKeroSurface(String id) {
	if ( surfaces.containsKey(id) )
	    return surfaces.get(id);
	else
	    return surfaces.get("10");
    }


    public Drawable getSurfaceDrawable(String id, Resources res) {
	return surfaces.get(id).getSurfaceDrawable(res);
    }
}

File: src/main/kotlin/com/cattailsw/nanidroid/SurfaceManager.kt
package com.cattailsw.nanidroid

import com.cattailsw.nanidroid.surface.ParsedSurfaceEntry

/**
 * Kotlin domain catalog for the shell surfaces installed for one ghost.
 *
 * It owns catalog and speaker-default selection. Parsed surface data crosses
 * from the shell parser to the Compose rendering pipeline through this catalog.
 */
class SurfaceManager(@Suppress("UNUSED_PARAMETER") ghostid: String) {
    private val surfaces = linkedMapOf<String, ShellSurface>()
    private val parsedEntries = linkedMapOf<String, List<ParsedSurfaceEntry>>()

    internal fun addSurface(id: String, surface: ShellSurface): Int {
        surfaces[id] = surface
        parsedEntries.remove(id)
        return surfaces.size
    }

    internal fun addParsedSurface(
        id: String,
        surface: ShellSurface,
        entries: List<ParsedSurfaceEntry>,
    ): Int {
        surfaces[id] = surface
        parsedEntries[id] = entries.toList()
        return surfaces.size
    }

    fun getParsedSurfaceEntries(id: String): List<ParsedSurfaceEntry> =
        parsedEntries[id].orEmpty()

    fun containsSurface(id: String): Boolean = id in surfaces

    fun getTotalSurfaceCount(): Int = surfaces.size

    fun getSurfaceKeys(): Set<String> = surfaces.keys

    /** Returns null for an unknown exact surface id. */
    fun getSurface(id: String): ShellSurface? = surfaces[id]

    fun getSakuraSurface(id: String): ShellSurface =
        surfaces[id] ?: surfaces[SAKURA_DEFAULT_ID] ?: nullSurface

    fun getKeroSurface(id: String): ShellSurface =
        surfaces[id] ?: surfaces[KERO_DEFAULT_ID] ?: nullSurface

    private companion object {
        const val SAKURA_DEFAULT_ID = "0"
        const val KERO_DEFAULT_ID = "10"
        val nullSurface = ShellSurface()
    }
}
